[Experiment] Vertical tabs

sangwoo108 opened the pull request:
This new feature is at experimental stage for now. So it's hidden
behind flag('Vertical tabs')

Implemented:
* Expand and collapse via a button on top of tabs.
* Vertical scroll area for tabs
* TabCards
* Tab reordering via drag-and-drop
* Tab detaching and attaching via drag-and-drop
* Shape customization

TBD:
* Switching tabs layout direction on runtime
  - need to plumb pref to many UI components.
* Title area in frame
  - Show title or move up toolbar
  - Moving up toolbar can take some time.
* Layout animation

![vertial_tabs](https://user-images.githubusercontent.com/5474642/174064114-d9f2db35-bb36-441f-87bb-3352fb24ce2b.gif)

<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves  https://github.com/brave/brave-browser/issues/23467


## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
- Can display tabs in the window
- Can switch between the tabs
- Can open new tabs
- Tabs can detach from and attach to the window

simonhong commented:
Crashed when right click over the empty area in the vertical tab.
<img width="353" alt="Screen Shot 2022-06-20 at 2 34 45 PM" src="https://user-images.githubusercontent.com/6786187/174531939-dce6f666-7e61-4931-a84f-81fedaccb5d0.png">

simonhong commented:
Can't drag window. maybe that should be handled via `Show title or move up toolbar` TBD item?

simonhong commented on browser/ui/views/frame/brave_tab_strip_region_container.cc:
```diff
@@ -0,0 +1,235 @@
+namespace {
+
+class ScrollHeaderView : public views::Button {
+ public:
```
Let's use `METADATA_HEADER` for all new view controls.

sangwoo108 replied:
done.

simonhong commented on browser/ui/views/frame/brave_tab_strip_region_container.cc:
```diff
@@ -0,0 +1,235 @@
+  }
+
+ private:
+  TabStrip* tab_strip_ = nullptr;
```
`raw_ptr`.

simonhong commented on chromium_src/chrome/browser/ui/views/frame/tab_strip_region_view.h:
```diff
@@ -0,0 +1,9 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#define BRAVE_UI_VIEWS_FRAME_TAB_STRIP_REGION_FRIEND \
```
With below, we don't need to add corresponding patch files.
```
#define FrameColorsChanged                   \
  UnUsed() {}                                \
  friend class BraveTabStripRegionContainer; \
  void FrameColorsChanged
#include "src/chrome/browser/ui/views/frame/tab_strip_region_view.h"
#undef FrameColorsChanged
```

sangwoo108 replied:
Cool! done.

simonhong commented on chromium_src/chrome/browser/ui/views/frame/browser_view_layout.cc:
```diff
@@ -0,0 +1,14 @@
+#include "brave/browser/ui/views/tabs/brave_vertical_tab_utils.h"
+
+#define BRAVE_SKIP_IF_VERTICAL_TAB \
+  if (ShouldShowVerticalTabs())    \
```
Maybe we can add `BraveBrowserViewLayout` and handle this via `BraveBrowserViewLayout::LayoutTabStripRegion()` without adding new patch?

simonhong reviewed (commented):
Let's check again new patch files.
Maybe some of them can be deleted with subclassing.
I think it's good time to get familiar with our overriding technique :)

sangwoo108 commented on browser/about_flags.cc:
```diff
@@ -460,6 +468,17 @@ const flags_ui::FeatureEntry::Choice kBraveSkusEnvChoices[] = {
+    {"brave-vertical-tabs", \
+    flag_descriptions::kBraveVerticalTabsName, \
+    flag_descriptions::kBraveVerticalTabsDescription, \
+    kOsDesktop, \
```
Done. kOsWin and only for nightly / dev.

sangwoo108 commented on browser/ui/views/frame/brave_browser_view.cc:
```diff
@@ -159,34 +160,47 @@ BraveBrowserView::BraveBrowserView(std::unique_ptr<Browser> browser)
                           base::Unretained(this)));
 #endif
 
+  const bool can_have_vertical_tabs = browser_->is_type_normal();
```
Yeah., something was missing here. Added IsVerticalTabsEnabled() here. 

As BraveTabRegion container takes `ShouldShowVerticalTabs()`  into account when calculating preferred size, BrowserViewLayout doesn't have to care about it.

sangwoo108 commented on chromium_src/chrome/browser/ui/views/frame/browser_view.cc:
```diff
@@ -5,12 +5,16 @@
 #define ToolbarView BraveToolbarView
 #define BrowserTabStripController BraveBrowserTabStripController
 #define TabStrip BraveTabStrip
+#define BRAVE_BROWSER_UI_VIEWS_FRAME_HIDE_HORIZONTAL_TAB \
```
done.

sangwoo108 commented on chromium_src/chrome/browser/ui/views/tabs/tab_container.cc:
```diff
@@ -0,0 +1,35 @@
+
+#include "brave/browser/ui/views/tabs/brave_vertical_tab_utils.h"
+
+#define BRAVE_UI_VIEWS_TABS_TAB_CONTAINER_CALCULATE_PREFERRED_SIZE_FOR_VERT_TABS \
```
done.

sangwoo108 commented:
Thank you for great tips. It seems our strategy is minimizing patch files.

simonhong commented on browser/ui/views/frame/brave_browser_view.cc:
```diff
@@ -159,34 +161,48 @@ BraveBrowserView::BraveBrowserView(std::unique_ptr<Browser> browser)
   sidebar_container_view_ = brave_contents_container->AddChildView(
       std::make_unique<SidebarContainerView>(GetBraveBrowser()));
+#endif  // BUILDFLAG(ENABLE_SIDEBAR)
+  vertical_tabs_container_ = brave_contents_container->AddChildView(
```
Or it'll be no-op after this initialization?

sangwoo108 replied:
Ooops, you're right. Let me check it again.

sangwoo108 replied:
changed to check nullity in contents layout manager.

simonhong commented on patches/chrome-browser-ui-views-tabs-tab_drag_controller.cc.patch:
```diff
@@ -0,0 +1,39 @@
+   last_move_screen_loc_ = start_point_in_screen_.x();
+   initial_tab_positions_ = source_context->GetTabXCoordinates();
+ 
++  BRAVE_TAB_DRAG_CONTROLLER_INIT_LOC_FOR_VERTICAL_TABS
```
nit: We want to minimize patch file size also. Let's remove empty line adding below and from other patch files also.

sangwoo108 replied:
done.

simonhong commented on chromium_src/chrome/browser/ui/views/tabs/tab_strip_layout_helper.h:
```diff
@@ -0,0 +1,20 @@
+#ifndef BRAVE_CHROMIUM_SRC_CHROME_BROWSER_UI_VIEWS_TABS_TAB_STRIP_LAYOUT_HELPER_H_
+#define BRAVE_CHROMIUM_SRC_CHROME_BROWSER_UI_VIEWS_TABS_TAB_STRIP_LAYOUT_HELPER_H_
+
+#define BRAVE_UI_VIEWS_TABS_TAB_STRIP_LAYOUT_HELPER_SET_USE_VERTICAL_TABS \
```
With below, we can avoid corresponding patch file :) 
```
#define UpdateIdealBounds                                                      \
  UnUsed() { return 0; }                                                       \
  void set_use_vertical_tabs(bool vertical) { use_vertical_tabs_ = vertical; } \
                                                                               \
 private:                                                                      \
  bool use_vertical_tabs_ = true;                                              \
                                                                               \
 public:                                                                       \
  int UpdateIdealBounds

#include "src/chrome/browser/ui/views/tabs/tab_strip_layout_helper.h"

#undef UpdateIdealBounds
```

sangwoo108 replied:
Cool! Done.

simonhong commented on browser/ui/tabs/features.h:
```diff
@@ -0,0 +1,13 @@
+#define BRAVE_BROWSER_UI_TABS_FEATURES_H_
+
+#include "base/feature_list.h"
+
```
Not changed yet ? :)

sangwoo108 replied:
Oh, sorry. I thought you were talking about forward declaration above.

sangwoo108 replied:
done

simonhong replied:
nit:  `Feature` is not necessary from `kBraveVerticalTabsFeature` as it's in `feature` namesapce.

sangwoo108 commented:
humm.. noplatform bot keeps failing due to patch files which I didn't touch here. What 's going on here..?

simonhong commented:
> humm.. noplatform bot keeps failing due to patch files which I didn't touch here. What 's going on here..?

Sometimes, this kind of unknonw error fixed with clean build.  Did re-run.

simonhong reviewed (dismissed):
++

sangwoo108 commented:
@brave/chromium-src-reviewers @brave/patch-reviewers Could you take a look at this please? :)

bbondy commented:
FYI I tried to test this out on macOS but it says the feature is not available on my platform in chrome://flags

sangwoo108 commented on browser/about_flags.cc:
```diff
@@ -460,6 +468,17 @@ const flags_ui::FeatureEntry::Choice kBraveSkusEnvChoices[] = {
+    {kBraveVerticalTabsFeatureInternalName,  \
+    flag_descriptions::kBraveVerticalTabsName, \
+    flag_descriptions::kBraveVerticalTabsDescription, \
+    kOsWin, \
```
If you want to try this feature on other platforms, add your platform here.

bsclifton replied:
Should we add macOS by default too?

sangwoo108 replied:
Yes we can but currently MacOS seems to have critical issues including crash. My mac isn't delivered yet so I can't deal with it for now. What do you think? should I add mac here?

DavidRBailey replied:
Please add MacOS for testing.

sangwoo108 replied:
Good news! I got my Mac and fixed reported problems. I think we can ship this on Mac too! 

<img width="814" alt="image" src="https://user-images.githubusercontent.com/5474642/181729671-00adb093-b84c-4892-b06b-63824aeb9b10.png">

simonhong commented:
> > FYI I tried to test this out on macOS but it says the feature is not available on my platform in chrome://flags
> 
> Yeah. I enabled this on Windows only. As Mac has different frame implementation, I should revisit it. But if you want to test locally, you can enable it on brave/about_flags.cc . I left a comment there. Thanks ;)

When I tested this PR on my macOS, there is some issue (crash or usablilty).
So, I recommended to start this feature with Windows only first.
When @sangwoo108 gets macOS machine, he can resolve this :)

sangwoo108 commented:
Now patch files I touched are gone. 😊

simonhong commented on chromium_src/chrome/browser/ui/views/tabs/tab_drag_controller.h:
```diff
@@ -0,0 +1,51 @@
+                    bool just_attached) override;
+};
+
+// using TabDragControllerBrave = TabDragController;
```
leftover?

sangwoo108 replied:
oops, removed

simonhong commented on chromium_src/chrome/browser/ui/views/tabs/tab_strip_layout_helper.cc:
```diff
@@ -0,0 +1,16 @@
+#include "brave/browser/ui/views/tabs/brave_tab_strip_layout.h"
+#include "chrome/browser/ui/views/tabs/tab_strip_layout.h"
+
+#define CalculateTabBounds                                                    \
```
who use this?

sangwoo108 replied:
https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/views/tabs/tab_strip_layout_helper.cc;l=275;drc=991d3fc291bb5ab890f9d6cd0ee0c230be3edb01

It's used in tab_strip_layout_helper.cc and it's the only occurrence in that file.

simonhong replied:
ah, I see 👍🏼

simonhong reviewed (approved):
++ 👍🏼

sangwoo108 commented:
Hi, @goodov ! As we talked, all patch files are gone now! PTAL when you can :)

sangwoo108 commented:
@goodov Friendly reminder :)

bridiver commented on browser/about_flags.cc:
```diff
@@ -42,6 +42,10 @@
 #include "brave/components/ipfs/features.h"
 #endif
 
+#if !BUILDFLAG(IS_ANDROID)
```
probably toolkit views is more appropriate here?

bridiver replied:
actually even better would be to add a buildflag for this and set it to `enable_vertical_tabs = toolkit_views` in gn

bridiver commented on browser/about_flags.cc:
```diff
@@ -422,6 +429,17 @@ constexpr char kRestrictWebSocketsPoolDescription[] =
 #define BRAVE_TRANSLATE_GO_FEATURE_ENTRIES
 #endif  // BUILDFLAG(ENABLE_BRAVE_TRANSLATE_GO)
 
+#if BUILDFLAG(IS_ANDROID)
```
same as above with TOOLKIT_VIEWS

bridiver commented on browser/about_flags.cc:
```diff
@@ -301,6 +305,9 @@ constexpr char kRestrictWebSocketsPoolName[] = "Restrict WebSockets pool";
 constexpr char kRestrictWebSocketsPoolDescription[] =
     "Limits simultaneous active WebSockets connections per eTLD+1";
 
+constexpr char kBraveVerticalTabsName[] = "Vertical tabs";
```
this should also have the buildflag

bridiver commented on browser/brave_features_internal_names.h:
```diff
@@ -7,5 +7,6 @@
 #define BRAVE_BROWSER_BRAVE_FEATURES_INTERNAL_NAMES_H_
 
 constexpr char kBraveVPNFeatureInternalName[] = "brave-vpn";
+constexpr char kBraveVerticalTabsFeatureInternalName[] = "brave-vertical-tabs";
```
buildflag

bridiver commented on browser/ui/views/tabs/brave_vertical_tab_utils.cc:
```diff
@@ -0,0 +1,18 @@
+  return base::FeatureList::IsEnabled(features::kBraveVerticalTabs);
+}
+
+bool ShouldShowVerticalTabs() {
```
why wouldn't you just call `IsVerticalTabsEnabled`?

bridiver replied:
also you should never create a global method like this with no namespace, but the problem is solved if you just call tabs::features::IsVerticalTabsEnabled

bridiver commented on browser/ui/views/tabs/brave_vertical_tab_utils.cc:
```diff
@@ -0,0 +1,18 @@
+#include "base/feature_list.h"
+#include "brave/browser/ui/tabs/features.h"
+
+bool IsVerticalTabsEnabled() {
```
this should move into features

bridiver commented on browser/ui/tabs/features.h:
```diff
@@ -0,0 +1,19 @@
+struct Feature;
+}  // namespace base
+
+namespace features {
```
this needs a top-level namespace `tabs::features::*`

sangwoo108 commented:
@bridiver Thank you for great comments! I addressed them. PTAL when you're ready :)

sangwoo108 commented:
Rebased onto the latest master.

bridiver commented on chromium_src/chrome/browser/unexpire_flags.cc:
```diff
@@ -18,7 +18,8 @@ namespace flags {
                    const char* internal_name) {
-#if BUILDFLAG(ENABLE_BRAVE_VPN) || BUILDFLAG(ENABLE_PLAYLIST)
+#if BUILDFLAG(ENABLE_BRAVE_VPN) || BUILDFLAG(ENABLE_PLAYLIST) || \
+    defined(TOOLKIT_VIEWS)
```
this doesn't seem correct, what are we trying to do here? Only enable this feature for certain channels? This doesn't seem like the right way to do that

sangwoo108 replied:
Yes, it's for controlling channel. I've tried https://github.com/brave/brave-core/pull/14198/files#r922671148 but was reverted due to CI failure. 

https://github.com/brave/brave-core/pull/14291

sangwoo108 replied:
Maybe we could have avoided CI failure of `AllowForDevVersion() - link  above` with `base::check_is_test`. What do you think about the `AllowForDevVersion()`? If you prefer this way, I can try it again.

bridiver commented on browser/ui/tabs/features.cc:
```diff
@@ -0,0 +1,21 @@
+namespace tabs {
+namespace features {
+
+const base::Feature kBraveVerticalTabs{"BraveVerticalTabs",
```
re https://github.com/brave/brave-core/pull/13816/files#r934749501 - it seems like we should enable/disable by channel here?

bridiver replied:
the closest thing I could find to this was `BorealisFeatures::MightBeAllowed` which would actually mean keeping `IsVerticalTabsEnabled`, but adding the channel check there

bridiver replied:
I'm not really sure what the best option is here. @iefremov @goodov any opinions? The expire thing seems weird to me.

goodov replied:
I prefer Griffin for this type of activity.

bridiver commented on browser/ui/tabs/features.cc:
```diff
@@ -0,0 +1,21 @@
+const base::Feature kBraveVerticalTabs{"BraveVerticalTabs",
+                                       base::FEATURE_DISABLED_BY_DEFAULT};
+
+bool IsVerticalTabsEnabled() {
```
personally I'm not a fan of these helper methods if they just check the feature value. It's easier to find things if you can just search for the flag name

sangwoo108 commented on browser/about_flags.cc:
```diff
@@ -479,6 +495,33 @@ constexpr char kPlaylistDescription[] = "Enables Playlist";
+    // During unit test, this can happen.
+    return 0;
+  }
+#endif
```
Yay, this seems to work. tested on test-release bot

sangwoo108 replied:
Now it looks working! - ci test-release bot passed. @bridiver What do you think about this? I like what you suggested from https://github.com/brave/brave-core/pull/13816/files#r934765872 but we might need this too in order to hide the flag from "brave://flags" page. I guess that's why folks have touched unexpire_flags.

bridiver replied:
this is only affecting the display of the flag, not the enable/disable state

sangwoo108 replied:
Yeah, that's intended. I didn't intend to change default "enable/disable" state per channels. Default state should be "disabled" for all channel before this feature get stable. And allowing to enable this feature only on Beta and Nightly is what we want here.

sangwoo108 replied:
@simonhong Do you think it's worth hiding this flag from stable channel? If we don't find a decent way to hide this, I think showing the flag is okay, as the default value is disabled.

bridiver replied:
I don't think trying to hide things from channels this way makes sense. Why would we hide the flag? We can control whether it's enabled or disabled by default through griffin and if people want to change the flag then they are volunteering to beta test it

simonhong replied:
same with @bridiver . If this is testable, it seems ok to displaying on all channels.

bridiver commented on browser/about_flags.cc:
```diff
@@ -486,6 +502,33 @@ constexpr char kPlaylistDescription[] = "Enables Playlist";
+#if defined(TOOLKIT_VIEWS)
+uint16_t AllowForDevVersion(uint16_t os) {
+#if BUILDFLAG(IS_WIN)
+  if (!install_static::InstallDetails::IsSet()) {
```
if this is happening in unit tests then you're missing setup that needs to happen for channel. I don't think this is the right way to fix it

bridiver commented on browser/about_flags.cc:
```diff
@@ -486,6 +502,33 @@ constexpr char kPlaylistDescription[] = "Enables Playlist";
+    {kBraveVerticalTabsFeatureInternalName,  \
+    flag_descriptions::kBraveVerticalTabsName, \
+    flag_descriptions::kBraveVerticalTabsDescription, \
+    AllowForDevVersion(kOsWin | kOsMac), \
```
this is gated on toolkit views everywhere else so we should conditionally include this based on toolkit views, not os. The channel portion will be handled by griffin

sangwoo108 commented:
![image](https://user-images.githubusercontent.com/5474642/183872005-8b6f851f-c5ef-43f0-a3aa-ba4b1c352d16.png)

I checked Linux too and enabled the flag on Linux too.

@bridiver Could you take a look again? I moved the feature under views/ like you said :)

bridiver commented on browser/ui/views/frame/brave_browser_view.cc:
```diff
@@ -172,34 +175,56 @@ BraveBrowserView::BraveBrowserView(std::unique_ptr<Browser> browser)
 
+  const bool can_have_vertical_tabs =
+      browser_->is_type_normal() &&
+      base::FeatureList::IsEnabled(tabs::features::kBraveVerticalTabs);
```
`ShouldShowVerticalTabs()` ?

bridiver commented on browser/ui/views/frame/brave_browser_view.cc:
```diff
@@ -172,34 +175,56 @@ BraveBrowserView::BraveBrowserView(std::unique_ptr<Browser> browser)
+  const bool can_have_vertical_tabs =
+      browser_->is_type_normal() &&
+      base::FeatureList::IsEnabled(tabs::features::kBraveVerticalTabs);
+  const bool can_have_sidebar = sidebar::CanUseSidebar(browser_.get());
```
`CanUseSidebar` needs to be guarded by `ENABLE_SIDEBAR`. This should be made more clear by adding `assert(enable_sidebar)` to `browser/ui/sidebar/BUILD.gn` but the include for sidebar_utils.h in this file is behind the flag so this needs to be as well.

bridiver commented on browser/ui/views/frame/brave_browser_view.cc:
```diff
@@ -172,34 +175,56 @@ BraveBrowserView::BraveBrowserView(std::unique_ptr<Browser> browser)
+      need_to_wrap_contents_container || can_have_sidebar;
+#endif  // BUILDFLAG(ENABLE_SIDEBAR)
+
+  if (!need_to_wrap_contents_container)
```
the buildflag guard for `ENABLE_SIDEBAR` should end after this return

bridiver replied:
actually ENABLE_SIDEBAR should just go away and sidebar should move into ui/views, but that's a pretty big change so let's just make sure the guards are correct in this PR

bridiver commented on browser/ui/views/frame/brave_tab_strip_region_container.cc:
```diff
@@ -0,0 +1,284 @@
+
+  ScrollHeaderView(Button::PressedCallback callback, TabStrip* tab_strip)
+      : Button(std::move(callback)), tab_strip_(tab_strip) {
+    // TODO(sangwoo.ko) Temporary workaround before we have a proper tooltip
```
please create an a tracking issue for this

sangwoo108 replied:
done.

bridiver commented on browser/ui/views/tabs/brave_tab_drag_context.cc:
```diff
@@ -0,0 +1,297 @@
+  // in that group.
+  int first_dragged_tab_index = group.has_value() ? 1 : 0;
+  if (static_cast<size_t>(first_dragged_tab_index) >= dragged_views.size()) {
+    // TODO(tbergquist): This shouldn't happen, but we're getting crashes
```
what is the crash stack trace? Can you replicate?

bridiver replied:
This code is copied from upstream? That's a major problem that needs to be resolved. What is different between BraveTabDragContext and TabDragContextImpl? We absolutely cannot have copied code here because it will get out of date and potentially be a problem even when vertical tabs are not enabled.

bridiver commented on chromium_src/chrome/browser/ui/views/tabs/tab_container.cc:
```diff
@@ -0,0 +1,57 @@
+
+  const int tab_overlap = TabStyle::GetTabOverlap();
+
+  // TODO(pkasting): When closing multiple tabs, we get repeated RemoveTabAt()
```
Please create a tracking issue for this, but we should follow the behavior from the normal tab strip

bridiver commented on browser/ui/views/frame/brave_browser_view.cc:
```diff
@@ -172,34 +175,56 @@ BraveBrowserView::BraveBrowserView(std::unique_ptr<Browser> browser)
   brave_contents_container->SetLayoutManager(
       std::make_unique<BraveContentsLayoutManager>(
-          sidebar_container_view_, original_contents_container_));
+          sidebar_container_view_, vertical_tabs_container_,
```
`sidebar_container_view_` is also behind `ENABLE_SIDEBAR` flag. This stuff is kind of a mess in here so it might be easier to do something like `enable_sidebar = toolkit_views` so we can drop the buildflags here as a first step towards https://github.com/brave/brave-core/pull/13816/files#r947320102

bridiver replied:
@simonhong in a separate PR can we address the ENABLE_SIDEBAR issue in general?

bridiver replied:
if something is specific to ui views, just put it in ui/views or provide better separation somehow so we don't have buildflag guards scattered around the code.

simonhong replied:
> @simonhong in a separate PR can we address the ENABLE_SIDEBAR issue in general?

Okay, I'll check if I can move another sidebar components to ui/views.

bridiver commented on browser/ui/views/frame/brave_tab_strip_region_container.h:
```diff
@@ -0,0 +1,51 @@
+}
+
+// Wraps TabStripRegion and show it vertically.
+class BraveTabStripRegionContainer : public views::View {
```
We shouldn't prepend `Brave` to something that is specific to vertical tab strip. These classes should be named based on vertical tab strip.

bridiver commented on browser/ui/views/frame/brave_tab_strip_region_container.h:
```diff
@@ -0,0 +1,51 @@
+class ScrollView;
+}
+
+// Wraps TabStripRegion and show it vertically.
```
it wraps TabStripRegionView, but why does it wrap it and not subclass?

bridiver commented on browser/ui/views/tabs/brave_tab_strip_layout.h:
```diff
@@ -0,0 +1,24 @@
+class Rect;
+}  // namespace gfx
+
+std::vector<gfx::Rect> CalculateVerticalTabBounds(
```
putting this in brave_tab_strip_layout.h further complicates the circular dependency issues and we should avoid that when possible. There's also no reason to expose this method publicly so I would put it in tab_strip_layout.cc chromium_src override

bridiver commented on browser/ui/views/tabs/brave_vertical_tab_utils.h:
```diff
@@ -0,0 +1,16 @@
+namespace tabs {
+
+// Returns true when users chose to use vertical tabs
+bool ShouldShowVerticalTabs();
```
this should go in features.h/.cc

sangwoo108 commented on browser/ui/views/frame/brave_browser_non_client_frame_view_mac.mm:
```diff
@@ -36,3 +37,12 @@
+
+int BraveBrowserNonClientFrameViewMac::GetTopInset(bool restored) const {
+  if (tabs::features::ShouldShowVerticalTabs()) {
+    // Set minimum top inset to show traffic lights on frame
```
Yeah, it's quite common nickname of window controls in MacOS. https://source.chromium.org/search?q=%22traffic%20light%22%20path:mac&sq=&ss=chromium

Is 'traffic light buttons' clearer? Or do you have any other suggestion?

sangwoo108 commented on chromium_src/chrome/browser/ui/views/frame/browser_frame_view_layout_linux.cc:
```diff
@@ -0,0 +1,17 @@
+
+int BrowserFrameViewLayoutLinux::NonClientTopHeight(bool restored) const {
+  if (tabs::features::ShouldShowVerticalTabs()) {
+    // Set minimum height to show caption buttons.
```
That's right. Similar reason.

goodov replied:
caption buttons seems clearer.

sangwoo108 replied:
Good.  replaced  traffic lights => caption buttons

sangwoo108 commented on browser/about_flags.cc:
```diff
@@ -504,6 +514,17 @@ constexpr char kPlaylistDescription[] = "Enables Playlist";
+    flag_descriptions::kBraveVerticalTabsName, \
+    flag_descriptions::kBraveVerticalTabsDescription, \
+    kOsWin | kOsMac | kOsLinux, \
+    FEATURE_VALUE_TYPE(tabs::features::kBraveVerticalTabs)},
```
Oh, yes. They should have one more space.

sangwoo108 commented on browser/ui/views/frame/brave_browser_view.h:
```diff
@@ -106,7 +100,8 @@ class BraveBrowserView : public BrowserView {
   raw_ptr<SidebarContainerView> sidebar_container_view_ = nullptr;
   raw_ptr<views::View> sidebar_host_view_ = nullptr;
-#endif
+
```
done.

sangwoo108 commented on browser/ui/views/tabs/brave_tab_drag_context.h:
```diff
@@ -0,0 +1,85 @@
+#include <memory>
+#include <vector>
+
+#include "chrome/browser/ui/tabs/tab_group.h"  // tab_drg_context.h missing this.
```
Removed.

Coffeeling commented:
@sangwoo108 Do these support tab groups, and if so, what do they look like?

sangwoo108 commented:
> @sangwoo108 Do these support tab groups, and if so, what do they look like?

Sorry, it's not considered here. It won't work well.

bridiver commented on components/sidebar/BUILD.gn:
```diff
@@ -6,6 +6,8 @@
 import("//brave/components/sidebar/buildflags/buildflags.gni")
 
 static_library("sidebar") {
+  assert(enable_sidebar)
```
also `enable_sidebar = is_win || is_mac || is_linux` -> `enable_sidebar = toolkit_views` in buildflags.gni

bridiver commented on browser/ui/views/frame/brave_browser_view.cc:
```diff
@@ -172,34 +171,50 @@ BraveBrowserView::BraveBrowserView(std::unique_ptr<Browser> browser)
-  if (!sidebar::CanUseSidebar(browser_.get())) {
+  const bool can_have_sidebar = sidebar::CanUseSidebar(browser_.get());
+
+  const bool need_to_wrap_contents_container =
```
`need_to_wrap_contents_container` is no longer needed

bridiver commented on browser/ui/views/frame/brave_browser_view.cc:
```diff
@@ -172,34 +171,50 @@ BraveBrowserView::BraveBrowserView(std::unique_ptr<Browser> browser)
 #endif
 
-#if BUILDFLAG(ENABLE_SIDEBAR)
+  const bool can_have_vertical_tabs =
```
nit: `can_have` makes it sound like it's just an option, but you're checking `ShouldShowVerticalTabs`. Maybe `show_vertical_tabs`? Same for sidebar

bridiver commented on browser/ui/views/frame/brave_browser_view.cc:
```diff
@@ -172,34 +171,50 @@ BraveBrowserView::BraveBrowserView(std::unique_ptr<Browser> browser)
   brave_contents_container->SetLayoutManager(
       std::make_unique<BraveContentsLayoutManager>(
-          sidebar_container_view_, original_contents_container_));
+          sidebar_container_view_, vertical_tabs_container_,
```
sidebar or vertical tabs might be null here because one does not imply the other

bridiver commented on chromium_src/chrome/browser/ui/views/tabs/tab_container.cc:
```diff
@@ -0,0 +1,50 @@
+
+// In order to instantiate RemoveTabDelegate which is defined in
+// tab_container.cc, this method is defined here.
+void BraveTabContainer::StartRemoveTabAnimation(Tab* tab,
```
same as BraveTabDragContext. We absolutely cannot have copied code from upstream here. This cannot break or cause any problems with regular tabs and copying code will almost certainly result in that happening over time.

bridiver commented on chromium_src/chrome/browser/ui/views/tabs/tab_drag_controller.cc:
```diff
@@ -0,0 +1,219 @@
+    return;
+  }
+
+  // Most of code here is borrowed from TabDragControllerChromium::Init() but we
```
same as above, we cannot copy code from upstream because it will get out of date and cause problems.

sangwoo108 reviewed (commented):
@bridiver Thank you so much. I removed most of copied code. There are 2 patch files I couldn't do with chromium_src/ override.

ryanbr commented:
Can the tab bar be resized? @sangwoo108

sangwoo108 commented:
> Can the tab bar be resized? @sangwoo108

Hi @ryanbr ! Do you mean resizing the width of the vertical tab bar by dragging? Not now, but in the task queue. This pr supports collapsing/expanding instead.

Creling commented:
Is it available in Brave Nightly? If not, is therer an estimate?

sangwoo108 commented:
> Is it available in Brave Nightly? If not, is therer an estimate?

Hi, @Creling . This PR is not merged yet so there's no available version at this moment. We're reviewing this code so that we can sustainably keep this feature. That's why it takes a bit longer. We definitely know how much users want this feature and trying our best. As for timeframe, this PR is merged as soon as it passes review process and follow-ups are scheduled.

> Does this have a color highlight/header somewhere to show what group the tabs are in and the ability to collapse?

Hi, @WHiZSTA . This PR doens't consider tab groups so it won't work well. But it's in task queue. We're trying to find out what's the best way to providing groups and other features in vertical tabs.

Creling commented:
Get it! Thank you for your effort. 😊

 

发件人: Sangwoo Ko ***@***.***> 
发送时间: 2022年8月30日 21:58
收件人: brave/brave-core ***@***.***>
抄送: Creling ***@***.***>; Mention ***@***.***>
主题: Re: [brave/brave-core] [Experiment] Vertical tabs (PR #13816)

 

Is it available in Brave Nightly? If not, is therer an estimate?

Hi, @Creling <https://github.com/Creling>  . This PR is not merged yet so there's no available version at this moment. We're reviewing this code so that we can sustainably keep this feature. That's why it takes a bit longer. We definitely know how much users want this feature and trying our best. As for timeframe, this PR is merged as soon as it passes review process and follow-ups are scheduled.

Does this have a color highlight/header somewhere to show what group the tabs are in and the ability to collapse?

Hi, @WHiZSTA <https://github.com/WHiZSTA>  . This PR doens't consider tab groups so it won't work well. But it's in task queue. We're trying to find out what's the best way to providing groups and other features in vertical tabs.

—
Reply to this email directly, view it on GitHub <https://github.com/brave/brave-core/pull/13816#issuecomment-1231708467> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKI4ZABH3G7ZY6GA57FC7HDV3YHP5ANCNFSM5Y6RQVDA> .
You are receiving this because you were mentioned.  <https://github.com/notifications/beacon/AKI4ZAFZC6O4SLTFXSJSWHTV3YHP5A5CNFSM5Y6RQVDKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOJFVGCMY.gif> Message ID: ***@***.*** ***@***.***> >

the4realpsych commented:
Hi @sangwoo108 , I've been following this thread for about a month now, and I'm very happy to see you and your teams dedication to providing vertical tabs in a soon-to-be update.

Just one question:
     - what's currently being built 
       into vertical tabs, and what's 
       the current progress?

Thank you very much; I'm just asking because I know that Brave will be THE STANDARD once vertical tabs are available - especially with the grand security features the browser provides. I'm a business person that's always on the web for work - so, I need to heavily-rely on a good browser for my business success.

sangwoo108 commented:
Hi, @the4realpsych ! Thank you for your interest :) Currently, we have only this PR so vertical tabs will be just like screen recording attached on top of this PR. And this PR is in review so that we can provide this feature sustainably in a long term. In the meantime, we're currently researching what vertical tabs should be like. Once we find out that, I think we can start improving this feature in earnest.

WHiZSTA commented:
Quick question, i was wondering if we could fast-track either this by utilizing existing code, or add a temporary alternative while this is being more fully developed.

I realized Brave already basically has vertical tabs coded, approved, and implemented in the browser.

That little downarrow next to the minimize button if you have ‘show tab search’ enabled under the brave->appearance settings.

We'd just need to:

1. make the tab search window dockable in the sidebar(or as a second sidebar) just add a pin button on the window to do so
2. shrink the size of the padding between items
3. remove the url + tab time and make them show in a tooltip instead
4. sort by default as actual tab order(so tab groups stay intact)
5. if you want yyou can provide a button for other sorts, like by group->time or group->title etc
6. move the recently closed list to a button next to the search so we dont have to scroll 500 items to get to it
7. reclaim/hide current title bar where tabs show and move min/max buttons down

sangwoo108 commented:
Hi, @WHiZSTA ! I'm so sorry that we're not fast enough to deliver this feature. But I can say that this is already in fast-track, because it was started without any approval or heavy process. We sincerely know how much our users eager this feature so we're trying our best.

> vertical tabs coded, approved, and implemented

Yes, we have one approval, but I need another approval from one of our core engineers. It's really necessary for us to be able to keep this feature in a long term. Unlike other features, when customizing existing Chromium code, this process can take a bit longer. I truly understand how long you've been waiting for this, and feel sorry about that. I can tell you that as soon as I get another approval, this PR will be merged and you can use it from the Nightly version.

Also, thank you for listing up the feature details. We're definitely considering them and I'll keep implementing those as experiments. Once again, thank you for bearing with us.

cdesouza-chromium commented on browser/ui/views/frame/brave_browser_non_client_frame_view_mac.mm:
```diff
@@ -38,3 +39,12 @@
+int BraveBrowserNonClientFrameViewMac::GetTopInset(bool restored) const {
+  if (tabs::features::ShouldShowVerticalTabs()) {
+    // Set minimum top inset to show caption buttons on frame.
+    return 18;
```
nit. Can this be a name constant?

the4realpsych commented:
> 

@sangwoo108 , is there any way to try the experimental features, or this still not available yet? Just asking out of curiosity. Thank you. ☺️

sangwoo108 commented:
> > 
> 
> @sangwoo108 , is there any way to try the experimental features, or this still not available yet? Just asking out of curiosity. Thank you. ☺️

Sorry, unless you can build Brave by yourself, there's no available version yet as far as I know. I sent a ping to our core engineers so I hope we can deliver this soon.

bridiver commented on browser/ui/views/frame/vertical_tab_strip_container.h:
```diff
@@ -0,0 +1,51 @@
+}
+
+// Wraps TabStripRegion and show it vertically.
+class VerticalTabStripContainer : public views::View {
```
this class is still not named according to convention. We don't really care that it wraps TabStripRegionView (not TabStripRegion) so I think this should be VerticalTabStripRegionView

sangwoo108 replied:
Yup. That looks better. Thanks.

bridiver commented on browser/ui/views/tabs/features.cc:
```diff
@@ -0,0 +1,22 @@
+                                       base::FEATURE_DISABLED_BY_DEFAULT};
+
+bool ShouldShowVerticalTabs() {
+  // TODO(sangwoo.ko) This should consider pref too.
```
please link to tracking issue

sangwoo108 replied:
Done.

bridiver commented on chromium_src/chrome/browser/ui/views/tabs/tab_drag_controller.cc:
```diff
@@ -0,0 +1,63 @@
+
+gfx::Point TabDragController::GetAttachedDragPoint(
+    const gfx::Point& point_in_screen) {
+  if (!tabs::features::ShouldShowVerticalTabs())
```
this doesn't look right, shouldn't you be calling the superclass method if ShouldShowVerticalTabs is not enabled?

bridiver commented on patches/chrome-browser-ui-views-tabs-tab_container.cc.patch:
```diff
@@ -0,0 +1,12 @@
+   // Compute the target bounds for animating this tab closed.  The tab's left
+   // edge should stay joined to the right edge of the previous tab, if any.
+   gfx::Rect target_bounds = tab->bounds();
++  TAB_CONTAINER_START_REMOVE_TAB_ANIMATION_FOR_VERTICAL_TABS
```
I think you need to rebase your branch, this doesn't match what I see in upstream source

bridiver commented on chromium_src/chrome/browser/ui/views/tabs/tab_strip.cc:
```diff
@@ -15,8 +16,35 @@
+    int y = 0;                                                            \
+    for (const TabSlotView* view : views) {                               \
+      const int height = view->height();                                  \
+      bounds.push_back(gfx::Rect(0, y, view->width(), view->height()));   \
```
`push_back` -> `emplace_back` and `view->height()` -> `height`

sangwoo108 replied:
Cool! Done.

bridiver commented on patches/chrome-browser-ui-views-tabs-tab_strip.cc.patch:
```diff
@@ -0,0 +1,20 @@
+ 
+   std::vector<gfx::Rect> CalculateBoundsForDraggedViews(
+       const std::vector<TabSlotView*>& views) override {
++    TAB_DRAG_CONTEXT_CALCULATE_BOUNDS_FOR_DRAGGED_VIEWS_VERTICAL_TABS
```
naming convention is `BRAVE_CALCULATE_BOUNDS_FOR_DRAGGED_VIEWS`

bridiver commented on patches/chrome-browser-ui-views-tabs-tab_strip.cc.patch:
```diff
@@ -0,0 +1,20 @@
+         continue;
+       }
+ 
++      TAB_DRAG_CONTEXT_CALCULATE_INSERTION_INDEX_FOR_VERTICAL_TABS
```
`BRAVE_CALCULATE_INSERTION_INDEX`

the4realpsych commented:
> > > 
> > 
> > 
> > @sangwoo108 , is there any way to try the experimental features, or this still not available yet? Just asking out of curiosity. Thank you. ☺️
> 
> Sorry, unless you can build Brave by yourself, there's no available version yet as far as I know. I sent a ping to our core engineers so I hope we can deliver this soon.

Okay, sounds good with me, @sangwoo108 ; very happy that your team has been so hard at work with this. I'm currently running the Brave beta, so I can't require to try vertical tabs out. Wish I could learn coding, though - just not sure where to start.

Thank you all for your dedication. ☺️

petemill commented:
I pushed a rebase @sangwoo108 so we can do a build with latest master

petemill commented on browser/about_flags.cc:
```diff
@@ -368,6 +373,11 @@ constexpr char kAllowCertainClientHintsDescription[] =
+#if defined(TOOLKIT_VIEWS)
+constexpr char kBraveVerticalTabsName[] = "Vertical tabs";
+constexpr char kBraveVerticalTabsDescription[] =
+    "Uses vertical tabs instead of tabstrip";
```
Well, I guess vertical or horizontal are still "strips" 😄 
Perhaps the description can be something like "Move tab strip to be a vertical panel on the side of the window instead of horizontal at the top of the window."?

sangwoo108 replied:
That's better. Thank!

petemill commented on browser/ui/views/frame/brave_browser_non_client_frame_view_mac.mm:
```diff
@@ -38,3 +39,12 @@
+int BraveBrowserNonClientFrameViewMac::GetTopInset(bool restored) const {
+  if (tabs::features::ShouldShowVerticalTabs()) {
+    // Set minimum top inset to show caption buttons on frame.
+    return 18;
```
Ordinarily I'd worry about hard-coding this. I'm not sure if macOS offers to make these buttons larger? I can't find anything. But either way this seems temporary since we'll add title area here or move toolbar up to this area.

sangwoo108 replied:
Yes, I agree. This is a temporary value just for this moment.

petemill commented on browser/ui/views/frame/vertical_tab_strip_region_view.cc:
```diff
@@ -0,0 +1,285 @@
+
+void VerticalTabStripRegionView::UpdateTabSearchButtonVisibility() {
+  const bool is_vertical_tabs = tabs::features::ShouldShowVerticalTabs();
+  if (auto* tab_search_button = region_view_->tab_search_button())
```
We will need to figure out where this goes as it is a useful feature.

sangwoo108 replied:
Yeah, on Windows, the search button will remain in frame area even with vertical tabs. This only happens on MacOS. I think vertical tabs' header area is a good place to put this button!

petemill reviewed (approved):
The code is a great basis for this feature. Thanks for working so hard on reducing the patches and overrides. It is much clearer and concise now too. Obviously when doing so much customization to the Tabs views it is worrying that we will have to rebase this on chromium and hope it won't be too painful. But for now, it's the right way and we're not ready to make a completely new tab strip implementation!

Along with your good list of TBD, and finishing the UI styles and transitions which we'll work with the design team on, here are some other issues that I think we can do in follow-ups:
- Dragging a tab downwards should not detach until it reaches the bottom of the tab strip container, i.e. the bottom of the window. Right now, it detaches past the NTP button which I think is not consistent with horizontal tabs.
- The vertical tab strip should (almost certainly) be to the side of the infobar and bookmarks bar. This is to avoid the "flash" of it's top moving up and down when switching between tabs which have those top bars and tabs which do not.
- Collapse state should be a pref which is used in other windows and persisted across browser restarts.
- Put the Tab Search button somewhere

petemill commented:
We need a test plan on this PR or on the Issue. Perhaps something simple, like steps to:
When turning on the flag
- Can display tabs in the window
- Can open new tabs

petemill commented:
I'm going to assume the minimal test coverage is ok whilst we're in this experimental form. When we define the feature more and add the UI polish then we can make more tests

sangwoo108 commented:
> I'm going to assume the minimal test coverage is ok whilst we're in this experimental form. When we define the feature more and add the UI polish then we can make more tests

Thank you @petemill ! Updated test plan section and rebased to resolve conflicts

sangwoo108 commented:
I really appreciate all reviewers for reviewing this for this long time, and users for waiting :)

Coffeeling commented:
Amazing to see this finally get merged. When can we expect to see the flag? 😍

sangwoo108 commented:
Hi, @Coffeeling! Nightly version with this is supposed to be built on next Monday 12pm(UTC). So you should be able to see it after some hours. And the next Beta version will have this flag soonish.

the4realpsych commented:
Hi @sangwoo108 , just read the comment above about the beta version having the brave://flag soon; will this be in the next beta version? Thank you very much, everyone, in advance. ☺️

sangwoo108 commented:
> Hi @sangwoo108 , just read the comment above about the beta version having the brave://flag soon; will this be in the next beta version? Thank you very much, everyone, in advance. ☺️

Hi @the4realpsych ! This feature is shipped on current Beta. If you don't see "Vertical tabs" in `brave://flags` page, please check out your version from brave://settings/help . It should be 1.45.xx .

the4realpsych commented:
> > Hi @sangwoo108 , just read the comment above about the beta version having the brave://flag soon; will this be in the next beta version? Thank you very much, everyone, in advance. ☺️
> 
> Hi @the4realpsych ! This feature is shipped on the current Beta. If you don't see "Vertical tabs" in `brave://flags` page, please check out your version from brave://settings/help . It should be 1.45.xx .

Absolutely wonderful, Sangwoo; does it support tab groups yet?

sangwoo108 commented:
> Absolutely wonderful, Sangwoo; does it support tab groups yet?

Sorry about that. tab groups are not working well for now

ghostlyentity commented:
Would be really cool for vertical tabs groups to be ironed out, would make it a serious contender for certain power users(myself as well of course).
Id love to see this task prioritized as i really cant do without such a feature (see others like firefox: sidebery, tree style tabs, treetabs, or chromium: forest, tabs outliner or tabli)

Coffeeling commented:
> > Absolutely wonderful, Sangwoo; does it support tab groups yet?
> 
> Sorry about that. tab groups are not working well for now. May I ask you how actively you use tab groups? I might be able to prioritize the task.

I use tab groups nonstop, impossible to handle the insane hoards of tabs otherwise.

ghostlyentity commented:
the most similar honestly just might be sidebery as it uses groups and not folder, perhaps look at it a little on the behavioral side of things?
in addition, which would be a really cool addition is to be able to zoom in the sidebar to make the tabs smaller or bigger.

thorrrr commented:
HI this is great news are you adding groups as well ?

JasonMorris1 commented:
Hi. Some features that would be nice is to have is the ability to close all tabs below a tab by right clicking on the tab and selecting the option. Simmilar to the feature on edge.

![image](https://user-images.githubusercontent.com/22126616/202363286-53eeec3c-a2f8-48ba-855f-eae9c9436c62.png)

thorrrr commented:
I ts the only killer feature and it was well implemented in Edge. But i have come home to Brave knowing this is being done.
For me just that feature kept me with the spying Edge :) I also like the minimise to icons in the tabs and groups

thorrrr commented:
KIll tabs below is the only thing  missing IMHO

thorrrr commented:
Hello i think the collapsed view need a little more work it looks a little to small on a high def screen and no padding.
Also the selected site its icon shows only a X not the site icon?
You then are not sure what site you are on?

![image](https://user-images.githubusercontent.com/8170673/202392841-c1b66453-7469-48ae-bc0e-f2b18e89d737.png)

thorrrr commented:
Sorry i added this [LINK](https://github.com/brave/brave-browser/issues/23467#issuecomment-1371971073)  in a closed post

owsiankakamila commented:
It works so well on Windows and I would **kill** for it to work as smoothly on Mac... because now it's a mess:

![image](https://i.imgur.com/PiVx5Ba.gif)
